kernel/src/snapshot.rs

```diff
@@ -321,29 +321,64 @@ fn read_last_checkpoint(
 
 /// List all log files after a given checkpoint.
 fn list_log_files_with_checkpoint(
-    cp: &CheckpointMetadata,
+    checkpoint_metadata: &CheckpointMetadata,
     fs_client: &dyn FileSystemClient,
     log_root: &Url,
 ) -> DeltaResult<(Vec<ParsedLogPath>, Vec<ParsedLogPath>)> {
-    let version_prefix = format!("{:020}", cp.version);
+    let version_prefix = format!("{:020}", checkpoint_metadata.version);
     let start_from = log_root.join(&version_prefix)?;
 
-    let files: Vec<ParsedLogPath> = fs_client
-        .list_from(&start_from)?
-        .flat_map(|file| file.and_then(ParsedLogPath::try_from).transpose())
-        .try_collect()?;
-    let (mut commit_files, checkpoint_files): (Vec<_>, _) = files
-        .into_iter()
+    let mut max_checkpoint_version = checkpoint_metadata.version;
+    let mut checkpoint_files = vec![];
+    let mut commit_files = Vec::with_capacity(10);
+
+    for meta_res in fs_client.list_from(&start_from)? {
+        let meta = meta_res?;
+        let parsed_path = ParsedLogPath::try_from(meta)?;
         // TODO this filters out .crc files etc which start with "." - how do we want to use these kind of files?
-        .filter(|p| p.is_commit() || p.is_checkpoint())
-        .partition(ParsedLogPath::is_commit);
+        if let Some(parsed_path) = parsed_path {
+            if parsed_path.is_commit() {
+                commit_files.push(parsed_path);
+            } else if parsed_path.is_checkpoint() {
+                match parsed_path.version.cmp(&max_checkpoint_version) {
+                    Ordering::Greater => {
+                        max_checkpoint_version = parsed_path.version;
+                        checkpoint_files.clear();
+                        checkpoint_files.push(parsed_path);
+                    }
+                    Ordering::Equal => checkpoint_files.push(parsed_path),
+                    Ordering::Less => {}
+                }
+            }
+        }
+    }
+
+    if checkpoint_files.is_empty() {
+        // TODO: We could potentially recover here
+        return Err(Error::generic(
+            "Had a _last_checkpoint hint but didn't find any checkpoints",
+        ));
+    }
+
+    if max_checkpoint_version != checkpoint_metadata.version {
+        warn!(
+            "_last_checkpoint hint is out of date. _last_checkpoint version: {}. Using actual most recent: {}",
+            checkpoint_metadata.version,
+            max_checkpoint_version
+        );
+        // we may need to drop some commits that are after the actual last checkpoint
+        commit_files.retain(|parsed_path| parsed_path.version > max_checkpoint_version);
+    } else if checkpoint_files.len() != checkpoint_metadata.parts.unwrap_or(1) as usize {
```

Comment:
Does this work?
```suggestion
    } else if checkpoint_files.len() != checkpoint_metadata.parts.unwrap_or(1usize) {
```
(replace `usize` with the suffix for whatever type `parts` has)?

Reply:
no. the issue is that `checkpoint_metadata.parts` was an i32, so the `1` was implicitly an i32, and trying to make it a `usize` doesn't work because the cast needs to happen _after_ the unwrap.

Anyway, since the reason we have `parts` is to compare it against a vector len, I made it a `usize` in the struct, and we don't have to cast. It means serde will fail if there's a negative number in the json, but that's _probably_ okay since that's a broken _last_checkpoint anyway.